### PR TITLE
custom auto increment id

### PR DESCRIPTION
**Pull request by nasr18**
Already using `mongoose-auto-increment` in our old project to auto-increment field. Now we are migrating our app to `nestjs` and `mongoose-auto-increment` broken due to typegoose. Hope this PR will fix that.

### CODE REVIEW

File: src/autoIncrement.ts [hunk 1]
```diff
@@ -54,26 +54,32 @@ export function AutoIncrementSimple(
 
 /** The Schema used for the trackers */
 const IDSchema = new mongoose.Schema({
-  f: String,
-  m: String,
-  c: Number
+  field: String,
```
Comment (hasezoey) [inline src/autoIncrement.ts]: Why change that to the full name?
-> it was smaller so that it requires less bandwidth and lower storage footprint

**Comment by hasezoey** [comment]
because you didnt give much information on what you did in this pull request, i will try to summarize it:
- expanded the names in the collection
- changed some names
- removed option `trackerCollection`
- changed default name for `trackerModelName` from `idtracker` to `identitycounter`
- added required option `model`
- added optional option `startAt`

File: src/autoIncrement.ts [hunk 2]
```diff
@@ -94,18 +100,28 @@ export function AutoIncrementID(schema: mongoose.Schema<any>, options: AutoIncre
 
     if (!model) {
       logger.info('Creating idtracker model named "%s"', opt.trackerModelName);
-      model = this.db.model(opt.trackerModelName, IDSchema, opt.trackerCollection);
+      model = this.db.model(opt.trackerModelName, IDSchema);
+      model.findOne({ model: opt.model, field: opt.field } as AutoIncrementIDTrackerSpec).lean()
+        .then((counter: AutoIncrementIDTrackerSpec) => {
+          if (!counter) {
+            model.create({
+              model: opt.model,
```
Comment (hasezoey) [inline src/autoIncrement.ts]: i didnt add this *this way* because it would be incorrect, so i used (later in the code) `this.constructor.modelName`

File: src/autoIncrement.ts [hunk 3]
```diff
@@ -94,18 +100,28 @@ export function AutoIncrementID(schema: mongoose.Schema<any>, options: AutoIncre
...
+              field: opt.field,
+              count: opt.startAt - opt.incrementBy
```
Comment (hasezoey) [inline src/autoIncrement.ts]: why decrement it?

File: src/types.ts [hunk 4]
```diff
@@ -10,17 +10,12 @@ export interface AutoIncrementOptionsSimple {
...
   /**
    * How much to increment the field by
    * @default 1
    */
   incrementBy?: number;
-  /**
-   * The Tracker Collection to use to keep track of an counter for the ID
-   * @default idtracker
-   */
-  trackerCollection?: string;
```
Comment (hasezoey) [inline src/types.ts]: why remove this option?

File: src/types.ts [hunk 5]
```diff
@@ -29,16 +24,25 @@ export interface AutoIncrementOptionsID {
   field?: string;
   /**
    * Set the tracker model name
-   * @default idtracker
+   * @default identitycounter
```
Comment (hasezoey) [inline src/types.ts]: why change the default name to this?

File: src/types.ts [hunk 6]
```diff
@@ -29,16 +24,25 @@ export interface AutoIncrementOptionsID {
...
+   * the count should start at
+   * @default 0
+   */
+  startAt?: number;
 }
 
 export interface AutoIncrementIDTrackerSpec {
   /** The ModelName from the current model */
-  m: string;
+  model: string;
```
Comment (hasezoey) [inline src/types.ts]: Why change that to the full name?
-> it was smaller so that it requires less bandwidth and lower storage footprint

(Note: deleted original, because it was in the wrong spot & and was an single comment, but should be in this review)

**Comment by nasr18** [comment]
> because you didnt give much information on what you did in this pull request, i will try to summarize it:
> 
>     * expanded the names in the collection
> 
>     * changed some names
> 
>     * removed option `trackerCollection`
> 
>     * changed default name for `trackerModelName` from `idtracker` to `identitycounter`
> 
>     * added required option `model`
> 
>     * added optional option `startAt`
> 
> 
> is this correct?

@hasezoey yes above mentioned points where correct.

File: src/autoIncrement.ts (hunk 3 above)
Comment (nasr18) [reply src/autoIncrement.ts]: if we want it to start from 10 means, we have to decrement by one and it will get stored in a separate collection with the model we have given. Then when we save any docs it will start count from 10.

File: src/types.ts (hunk 4 above)
Comment (nasr18) [reply src/types.ts]: we let mongoose to apply their way of naming the collections. they will lowercase and pluralize the collection name from model name.

File: src/types.ts (hunk 5 above)
Comment (nasr18) [reply src/types.ts]: because the plugin `mongoose-auto-increment` where using this name. and I dont wanna change that.

File: src/types.ts (hunk 6 above)
Comment (nasr18) [reply src/types.ts]: it should be readable. that's why changed to full name.

File: src/autoIncrement.ts (hunk 3 above)
Comment (hasezoey) [reply src/autoIncrement.ts]: right forgot about the $inc

File: src/types.ts (hunk 4 above)
Comment (hasezoey) [reply src/types.ts]: but why not give the user the option to change it to something that suits to the style of naming things, including disabling pluralization

File: src/types.ts (hunk 5 above)
Comment (hasezoey) [reply src/types.ts]: ok

File: src/types.ts (hunk 6 above)
Comment (hasezoey) [reply src/types.ts]: its fine by me, but just thought that it should require as low as possible
-> maybe (when asked to) implement an model that makes it "short"?

File: src/types.ts (hunk 4 above)
Comment (nasr18) [reply src/types.ts]: ok will revert this change.

File: src/autoIncrement.ts (hunk 2 above)
Comment (nasr18) [reply src/autoIncrement.ts]: we are using multiple models to use auto-increment id. so we have to mention that here.

Comment (hasezoey) [reply src/autoIncrement.ts]: with that i mean it should be got by the actual model later, and not induce some "user error" by inputting the name manually

Comment (nasr18) [reply src/autoIncrement.ts]: I understand what you are saying. but the plugin `mongoose-auto-increment` where forced us to enter model name and we did that wherever we need auto-increment id with different name. I mean, collection name and the model name will be different in that `identitycounters` collection.

File: src/autoIncrement.ts [hunk 7]
```diff
@@ -95,17 +102,27 @@ export function AutoIncrementID(schema: mongoose.Schema<any>, options: AutoIncre
     if (!model) {
       logger.info('Creating idtracker model named "%s"', opt.trackerModelName);
       model = this.db.model(opt.trackerModelName, IDSchema, opt.trackerCollection);
+      model.findOne({ model: opt.model, field: opt.field } as AutoIncrementIDTrackerSpec).lean()
+        .then((counter: AutoIncrementIDTrackerSpec) => {
+          if (!counter) {
+            model.create({
+              model: opt.model,
```
Comment (hasezoey) [inline src/autoIncrement.ts]:
```suggestion
              model: (this.constructor as any).modelName,
```
this would be more dynamic and automatically gets the model name from the model the plugin was applied to, the only limitation i could find, was that no different connection than the parent could be used

Comment (nasr18) [reply src/autoIncrement.ts]: I agree with you. but the output of `getName(SomeClass)` is not as same as `mongoose collection` name. so we are getting duplicate entries in `identitycounters` collection.

Comment (hasezoey) [reply src/autoIncrement.ts]: sorry i dont understand you, this plugin dosnt use the collection-name of the applied model, only the modelname

Comment (nasr18) [reply src/autoIncrement.ts]: actually we where following `mongoose's full lowercase and pluralize collection name` everywhere. so we are facing trouble. that's why we need this extra field. I fixed this by making the model field optional.

Comment (hasezoey) [reply src/autoIncrement.ts]: could you give me a link to that, and explain why you want to store the collection name instead of the modelName?

PS: `getName(SomeClass)` is only meant when used with typegoose and its classes

File: src/autoIncrement.ts [hunk 8]
```diff
@@ -92,20 +94,31 @@ export function AutoIncrementID(schema: mongoose.Schema<any>, options: AutoIncre
...
       logger.info('Creating idtracker model named "%s"', opt.trackerModelName);
       model = this.db.model(opt.trackerModelName, IDSchema, opt.trackerCollection);
+      const counter: AutoIncrementIDTrackerSpec = await model.findOne({ model: modelName, field: opt.field }).lean().exec();
+      if (!counter) {
+        await model.create({
+          model: modelName,
+          field: opt.field,
+          count: opt.startAt > 0 ? opt.startAt - opt.incrementBy : 0
+        } as AutoIncrementIDTrackerSpec);
+      }
```
Comment (hasezoey) [inline src/autoIncrement.ts]: could you add an test for this? (that runs it once, checks, runs again to expect an error ("Duplicate")?

File: src/autoIncrement.ts (hunk 7 above)
Comment (nasr18) [reply src/autoIncrement.ts]: https://mongoosejs.com/docs/models.html#compiling

Comment (nasr18) [reply src/autoIncrement.ts]: our model name is `user` and same is given in `identitycounters model` field. but our class name starts with caps `class User` and so `getName(User)` returning `User` and duplicate entries got inserted as `model: 'User'`

File: src/autoIncrement.ts (hunk 8 above)
Comment (nasr18) [reply src/autoIncrement.ts]: its working fine. no `Duplicate` error.

Comment (hasezoey) [reply src/autoIncrement.ts]: > its working fine. no Duplicate error.

i meant to ask if you could add an test for this new (separate) "query"

File: src/autoIncrement.ts (hunk 7 above)
Comment (hasezoey) [reply src/autoIncrement.ts]: @nasr18, when you use typegoose's `getModelForClass` and dont use an custom name, then the model name is the same as the class, but the collection is still lowercased & pluralized

easy example:
```ts
// NodeJS: 12.15.0
// MongoDB: 4.2-bionic (Docker)
import { getModelForClass, prop } from "@typegoose/typegoose"; // @typegoose/typegoose@6.3.1
import * as mongoose from "mongoose"; // mongoose@5.8.11
import { getName } from "@typegoose/typegoose/lib/internal/utils";

class User {
  @prop()
  public username?: string;
}
const UserModel = getModelForClass(User);

(async () => {
  await mongoose.connect(`mongodb://localhost:27017/`, { useNewUrlParser: true, dbName: "verifyMASTER", useCreateIndex: true, useUnifiedTopology: true });

  console.log("User model name:", getName(User), UserModel.modelName); // log: User model name: User User
  console.log("User collection name:", UserModel.collection.name); // log: User collection name: users

  await mongoose.disconnect();
})();

```

and if you use some custom name (or automaticName turned on), then it will be the custom name and `getName` will return the correct one

but if you set the `collection` option, it dosnt affect the name, the collection name gets automaticly set based on the modelname (lowercased & pluralized) when no `collection` options is provided:
```ts
// NodeJS: 12.15.0
// MongoDB: 4.2-bionic (Docker)
import { getModelForClass, prop, modelOptions } from "@typegoose/typegoose"; // @typegoose/typegoose@6.3.1
import * as mongoose from "mongoose"; // mongoose@5.8.11
import { getName } from "@typegoose/typegoose/lib/internal/utils";

@modelOptions({ schemaOptions: { collection: "SomeCustomCollectionName" } })
class User {
  @prop()
  public username?: string;
}
const UserModel = getModelForClass(User);

(async () => {
  await mongoose.connect(`mongodb://localhost:27017/`, { useNewUrlParser: true, dbName: "verifyMASTER", useCreateIndex: true, useUnifiedTopology: true });

  console.log("User model name:", getName(User), UserModel.modelName); // log: User model name: User User
  console.log("User collection name:", UserModel.collection.name); // log: User collection name: SomeCustomCollectionName

  await mongoose.disconnect();
})();
```

that is why `this.constructor.modelName` is simpler and dosnt require user input

TL;DR: this plugin, `typegoose-auto-increment`'s function for `ID`'s is based on the *model name* and not on the *collection name*

Comment (nasr18) [reply src/autoIncrement.ts]: instead of passing `model` in `plugin`, passing `@modelOptions({ options: { customName: 'user' } })` fixed the problem.

Comment (hasezoey) [reply src/autoIncrement.ts]: @nasr18, good to hear that you issue got fixed

File: src/autoIncrement.ts [hunk 9]
```diff
@@ -94,7 +93,7 @@ export function AutoIncrementID(schema: mongoose.Schema<any>, options: AutoIncre
       return;
     }
 
-    const modelName = opt.model || (this.constructor as any).modelName;
+    const modelName = (this.constructor as any).modelName;
```
Comment (hasezoey) [inline src/autoIncrement.ts]:
```suggestion
    const modelName: string = (this.constructor as any).modelName;
```
just to not have an any type

File: src/autoIncrement.ts (hunk 8 above)
Comment (hasezoey) [reply src/autoIncrement.ts]: as a note, i will add one later when this is merged

File: src/autoIncrement.ts [hunk 10]
```diff
@@ -92,20 +93,31 @@ export function AutoIncrementID(schema: mongoose.Schema<any>, options: AutoIncre
...
+
     if (!model) {
       logger.info('Creating idtracker model named "%s"', opt.trackerModelName);
       model = this.db.model(opt.trackerModelName, IDSchema, opt.trackerCollection);
+      const counter: AutoIncrementIDTrackerSpec = await model.findOne({ model: modelName, field: opt.field }).lean().exec();
+      if (!counter) {
+        await model.create({
+          model: modelName,
+          field: opt.field,
+          count: opt.startAt > 0 ? opt.startAt - opt.incrementBy : 0
```
Comment (hasezoey) [inline src/autoIncrement.ts]: not allowing negative starting numbers?

File: src/autoIncrement.ts [hunk 11]
```diff
@@ -92,20 +93,31 @@ export function AutoIncrementID(schema: mongoose.Schema<any>, options: AutoIncre
       return;
     }
 
+    const modelName = (this.constructor as any).modelName;
+    logger.info('modelName "%s"', modelName);
```
Comment (hasezoey) [inline src/autoIncrement.ts]: maybe remove this debug log again, or make it have more sense

Comment (nasr18) [reply src/autoIncrement.ts]: will remove

File: src/autoIncrement.ts (hunk 10 above)
Comment (nasr18) [reply src/autoIncrement.ts]: not. do i have to?

Comment (hasezoey) [reply src/autoIncrement.ts]: i thought your earlier solution was fine, and just set the default `startAt` to `0`, but allow any number

Comment (nasr18) [reply src/autoIncrement.ts]: done

**Comment by hasezoey** [comment]
to make the tests working again, the expected value would have to be changed accordingly

**Comment by nasr18** [comment]
> to make the tests working again, the expected value would have to be changed accordingly
> 
> otherwise, i dont see any other changes necessary (except maybe later documentation)

@hasezoey test cases updated with different inputs.
